db/view/view.cc

```diff
@@ -1039,7 +1039,13 @@ future<> mutate_MV(
                 cf_stats.total_view_updates_failed_local += is_local;
                 cf_stats.total_view_updates_failed_remote += remotes;
                 auto ep = f.get_exception();
-                vlogger.error("Error applying view update to {}: {}", target, ep);
+                try {
+                    std::rethrow_exception(ep);
+                } catch (const exceptions::mutation_unstarted_write_cancelled_exception& e) {
+                    vlogger.debug("Error applying view update to {}: {}", target, e.what());
```

Comment:
Can the exception happen in other scenarios?

Perhaps the condition should be on the phase of execution rather than the type of the exception. If the code here receives an exception while shutting down, it can swallow it, rather than looking at its type. Or maybe this is fine, I can't really tell.

Reply:
Avi, the thing is that these "unstarted write cancelled" thing is not about shutdown at all. We cancel these writes when we realize the **remote** destination node is down. So it's not that *this* node is shutting down, it's other nodes which are shutting down - so this node doesn't know it is in this "phase of execution".
But I agree with your sentiment... There's nothing particularly interesting about this specific case and its exception. In fact, in my original patch, I didn't even throw an exception in this case - nobody cares if the view updated started or cancelled before it started. It's not an error or even unexpected for the write to be cancelled - it's our normal code path when a remote node shuts down.

Reply:
> I spent a while verifying and considering if this really should be an exception, and it looks like @nyh's original idea that it should not be an exception is more legitimate. In all call sites, it's handled fine when a regular, non-exceptional future is returned for a cancelled write - my initial suspicion was that it would be problematic, but it's not. So, logging a debug message and returning a ready future looks like the cleanest way to cancel an unstarted write, which is not really an exceptional state, because it happens naturally as a result of our `on_down()` handling.

I find it odd to return a non-exceptional future in such case. One could assume that if the write operation completes without exception it means that it was replicated. It doesn't matter that `on_down()` is a normal situation, the write operation did not succeed.

Why is logging of failed writes important when nodes are not going down, and it is not important when they are?

Reply:
There are two conflicting definitions of `write failed` here, so we need to pick one before approaching this patch... one is that if a write operation did not led to replicating data to other nodes, then it's considered failed, which suggests that it should result in an exceptional future. The other is that cancelling a yet unstarted write is not an error, but a regular maintenance rollback operation, so it doesn't deserve informing the client about the error or logging it with error severity, because from our side it's expected behaviour.

The earlier version of this pull request assumed the first approach (i.e. cancelling an unstarted write is a failure), while the current version assumes the second one. I considered it for a while and the second definition makes more sense to me, because cancelling a write that hasn't started yet is expected behaviour.

There's also a third option of leaving the semantics as is and amending the dtest to ignore this `error` in the logs, as it's not really an error, but an internal detail of how we handle the event of other nodes being down - I don't like this  option. The fourth option is to make sure that we never have to cancel an unstarted write in the first place, which will be elegant, but I imagine it would also involve making more operations atomic (so there's no preemption point between creating a response handler and `mutate_begin`), which may be rather complex to test and prove correct - also, this one can perhaps have performance side effects.